// cyclejs-storage/build/index.js
"use strict";
var xstream_adapter_1 = require('@cycle/xstream-adapter');
function adapt(storage, write$, runSA, extractor) {
    return runSA.adapt(write$.map(function (p) { return p.value; }).startWith(extractor(storage)), xstream_adapter_1.default.streamSubscribe);
}
function specificStorageDriver(storage, write$, runSA) {
    write$.addListener({
        next: function (pair) { return storage.setItem(pair.key, pair.value); },
        error: function () { },
        complete: function () { }
    });
    return {
        getItem: function (key) { return adapt(storage, write$, runSA, function (s) { return s.getItem(key); }); },
        nthKey: function (n) { return adapt(storage, write$, runSA, function (s) { return s.key(n); }); },
    };
}
function storageDriver(write$, runSA) {
    return {
        session: specificStorageDriver(sessionStorage, write$.filter(function (pair) { return pair.target === 'session'; }), runSA),
        local: specificStorageDriver(localStorage, write$.filter(function (pair) { return pair.target === 'local'; }), runSA)
    };
}
function bindAdapter(f) {
    f.streamAdapter = xstream_adapter_1.default;
    return f;
}
function makeLocalStorageDriver() {
    return bindAdapter(specificStorageDriver.bind(undefined, localStorage));
}
exports.makeLocalStorageDriver = makeLocalStorageDriver;
function makeSessionStorageDriver() {
    return bindAdapter(specificStorageDriver.bind(undefined, sessionStorage));
}
exports.makeSessionStorageDriver = makeSessionStorageDriver;
function makeStorageDriver() {
    return bindAdapter(storageDriver);
}
exports.makeStorageDriver = makeStorageDriver;
